[Work in progress] First draft for an algebraic graphs proposal

anfelor opened the pull request:
I am a student at the university of Bonn and applied (unsuccessfully) for last year's algebraic graphs proposal. This year, @snowleopard approached me and asked me to submit a proposal myself which you find below. It is (obviously) tailored to myself and contains the areas I would want to be working on while setting (I hope) reasonable goals.

Questions:
  1. What about the scope of the project? Would you consider the demands easy to fufilll or to harsh?
  2. Should the optional goals remain optional or maybe be included as required?
  3. @snowleopard Are you comfortable with the vision I have laid out for alga? Would you be willing to merge a not insignificant addition of code into alga, or should this proposal maybe be centered on providing an auxiliary library?

I am looking forward to your suggestions.

anfelor committed:
First draft for algebraic graphs proposal

snowleopard commented on content/ideas/algebraic-graphs.md:
```diff
@@ -0,0 +1,62 @@
+While graphs are ubiquitous in computer programming,
+they have long been underrepresented in Haskell
+due to the imperative nature and lack of type-safety
+of the existing implementations.
```
This is a bit harsh towards the Functional Graph Library `fgl`, which provides a purely functional approach to modelling and manipulating graphs. I would say that the design space is not yet fully explored, and new approaches are being suggested by the community, such as algebraic graphs, which is the focus of this project.

snowleopard commented on content/ideas/algebraic-graphs.md:
```diff
@@ -0,0 +1,62 @@
+due to the imperative nature and lack of type-safety
+of the existing implementations.
+
+The [Algebraic Graphs library](https://github.com/snowleopard/alga) has so far succeded 
```
Typo: "succeded" -> "succeeded"

snowleopard commented on content/ideas/algebraic-graphs.md:
```diff
@@ -0,0 +1,62 @@
+benefit downstream packages like `build` or `aura`.
+
+In particular, the result should be a type-safe abstraction,
+that makes it easier to work with algorithms like `scc` or `topSort`.
```
I think it would be useful to refer to existing GitHub issues, which are raised by the users of the Alga library.

snowleopard commented on content/ideas/algebraic-graphs.md:
```diff
@@ -0,0 +1,62 @@
+
+Acyclic graphs are both common and heavily used in dependency
+management. Improvements in this area would therefore directly
+benefit downstream packages like `build` or `aura`.
```
You can also link to IOHK's Plutus project, which uses Alga: 

https://github.com/input-output-hk/plutus/blob/master/plutus-ir/src/Language/PlutusIR/Analysis/Dependencies.hs

Many of Alga's issues appear to come from there.

snowleopard replied:
I'd also add something like ", as well as a few commercial users of the library" -- I know for sure that a few companies are using Alga, but unfortunately can't go into any specifics, as this information is not public yet.

snowleopard commented on content/ideas/algebraic-graphs.md:
```diff
@@ -0,0 +1,62 @@
+
+## Goal 2: Implement common graph algorithms like Kruskal, Dijkstra and Moore-Bellman-Ford
+
+Thanks to the edge-labels developed in the build-up of
```
"edge-labels" -> "edge-labelled algebraic graphs"

snowleopard commented on content/ideas/algebraic-graphs.md:
```diff
@@ -0,0 +1,62 @@
+as there are nearly zero algorithms provided.
+
+The student should provide algorithms solving the following problems 
+on at least one representation:
```
I'd drop "on at least one representation" as it's not immediately clear what this means.

snowleopard commented:
@anfelor Thanks! I've left a few comments in the text.

> What about the scope of the project? Would you consider the demands easy to fufilll or to harsh?

I think the scope is good, but @jaspervdj might be in a better position to comment.

> Should the optional goals remain optional or maybe be included as required?

I really like the second optional goal, and I think it should actually be merged into the first two goals. The output is going to be not just code, but also tutorials/blog posts/functional pearls -- they will explain key  ideas behind the implementation and how to use new functionality. 

The maximum flow optional goal feels unrealistic to me. There is a lot of work in the first two goals already, so I'd just remove it.

> @snowleopard Are you comfortable with the vision I have laid out for alga? Would you be willing to merge a not insignificant addition of code into alga, or should this proposal maybe be centered on providing an auxiliary library?

Both acyclic graphs, as well as classic graph algorithms, would be great additions to Alga. 

By the way, there is some non-trivial interaction between the two goals that you might mention. Shortest path finding is simpler and faster on acyclic graphs: one can simply traverse all vertices in the topological order, starting from the source vertex. This justifies the order of the two goals, and also emphasises the importance of being able to capture graph acyclicity in types.

anfelor commented:
> This is a bit harsh towards the Functional Graph Library fgl, which provides a purely functional approach to modelling and manipulating graphs.

That's right, I have deleted that paragraph for the time being.

> I really like the second optional goal, and I think it should actually be merged into the first two goals. The output is going to be not just code, but also tutorials/blog posts/functional pearls -- they will explain key ideas behind the implementation and how to use new functionality.

I agree! But what should the measurement for successful completion be?
Or maybe I should just add "and document these in a blog post or article" a few times?

> The maximum flow optional goal feels unrealistic to me. There is a lot of work in the first two goals already, so I'd just remove it.

I have already implemented all of the above algorithms in C/C++ and even though finding efficient and clean Haskell implementations will take some time, I could imagine that there might be some time left towards the end. That's why I marked it as optional; not necessary for successful completion of the GSoC but a possible direction once the other goals are achieved.

> Both acyclic graphs, as well as classic graph algorithms, would be great additions to Alga.

Yay!

anfelor committed:
Changes as discussed in #101

snowleopard commented:
> I agree! But what should the measurement for successful completion be?
Or maybe I should just add "and document these in a blog post or article" a few times?

You could just add a short paragraph after the list of goals, specifying what it means for a goal to be reached, e.g. corresponding PRs merged, and documented in a blog post or article.

> I have already implemented all of the above algorithms in C/C++ and even though finding efficient and clean Haskell implementations will take some time, I could imagine that there might be some time left towards the end. That's why I marked it as optional; not necessary for successful completion of the GSoC but a possible direction once the other goals are achieved.

OK, sounds good!

Let me also ping @nobrakal who currently helps maintain Alga and was interested in helping out as a mentor. Perhaps he has some other comments.

snowleopard commented on content/ideas/algebraic-graphs.md:
```diff
@@ -0,0 +1,61 @@
+in part due to the lack of graph algorithms provided.
+
+Therefore the goal of this year's Summer of Code is
+to increase the usefulness of alga for a wider audience:
```
Let's capitalise Alga consistently throughout the proposal.

nobrakal commented:
Hi there,

This proposal sounds good for me :) There is indeed a lot of (really interesting) work for the first two goals (implementing each algorithm is I think a challenge).

As noticed by @snowleopard, I would switch the order of the two optional goals, the second being really interesting itself and allowing communication with the community (as I experienced, people are really interested by GSoC projects and are eager to understand what is done, how it is done and why ^^).

jaspervdj commented on content/ideas/algebraic-graphs.md:
```diff
@@ -0,0 +1,61 @@
+The [Algebraic Graphs library](https://github.com/snowleopard/alga) has so far succeeded
+(thanks in part to last years summer of code) in providing a common, 
+typesafe interface for graph instances. However, general ecosystem adoption remains low,
+in part due to the lack of graph algorithms provided.
```
I would replace the last sentence by "However, a lot of graph algorithms are still missing".

jaspervdj commented on content/ideas/algebraic-graphs.md:
```diff
@@ -0,0 +1,61 @@
+distances in Alga; however, this feature is rarely used
+as there are nearly zero algorithms provided.
+
+The student should provide algorithms solving the following problems:
```
I would put a concrete number on this, e.g. 5 algorithms, and I guess this is obvious but also require the corresponding tests

jaspervdj commented on content/ideas/algebraic-graphs.md:
```diff
@@ -0,0 +1,61 @@
+
+In particular, one could take problems that are unconnected to alga
+and show (maybe in the format of a functional programming pearl) how to
+solve them with alga.
```
I think this optional goal is fine, but keep in mind that according to the GSoC guidelines the project focus must be on code/implementation rather than documentation

jaspervdj reviewed (commented):
@anfelor Thanks for writing this up!  I think this proposal looks decent, let me know when you want me to merge this.

anfelor committed:
Further changes as discussed in #101

anfelor commented:
> I think this proposal looks decent, let me know when you want me to merge this.

I am happy with the proposal as it currently is; if @snowleopard and @nobrakal are too, this is ready to be merged.

> I really like the second optional goal, and I think it should actually be merged into the first two goals.

I have made adding tests and documentation part of the first two goals and removed the 2. optional goal accordingly.

Do you think the criteria for success are clear enough in this version or should I specify them a bit more?

snowleopard commented on content/ideas/algebraic-graphs.md:
```diff
@@ -24,35 +23,34 @@ that makes it easier to work with algorithms like `scc` or `topSort`
+ - Finding shortest paths in a graph
+   - with positive edge-weigths (Dijkstra)
+   - with conservative edge-weights (Moore-Bellman-Ford)
+   - between all vertices (Floyd-Warshall or Dijkstra+MBF)
```
I'd like to point out that Alga already contains a rather straightforward implementation of the Warshall-Floyd-Kleene algorithm, which, for an appropriate choice of the underlying semiring, gives you a shortest paths:

https://hackage.haskell.org/package/algebraic-graphs-0.3/docs/Algebra-Graph-Labelled-AdjacencyMap.html#v:closure

I'm sure the current implementation can be improved, but it may be useful to mention this in the proposal, as a baseline implementation, with with you will be comparing your results.

snowleopard commented:
@anfelor Looks good. I've added one more comment, which should help you make success criteria a bit more concrete: the path-finding algorithms you develop should be an improvement over the existing unoptimised and overly generic implementation of the `closure` operation.

Also, it may be worth comparing the resulting performance to `fgl`, which does contain most of the algorithms you list. To do that, you can utilise the benchmarking framework developed in the previous GSoC.

nobrakal commented:
> @nobrakal are too, this is ready to be merged.

Good for me :)

> Also, it may be worth comparing the resulting performance to fgl, which does contain most of the algorithms you list. To do that, you can utilise the benchmarking framework developed in the previous GSoC.

Indeed, and it will be a good reason for me to keep working on it ^^
